File: src/com/train/algorithm/tree/implementInC++/BinaryTreeMaximumPathSum.cpp
//
// Created by 罗斯 on 9/11/2022.
//
#include <algorithm>

using namespace std;
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    int maxPathSum(TreeNode* root) {
        if (!root) return 0;
        int res = INT_MIN;
        maxPathSum(root, res);
        return res;
    }
private:
    int maxPathSum(TreeNode* root, int& res) {
        if (!root) return 0;
        int l = max(maxPathSum(root->left, res), 0);
        int r = max(maxPathSum(root->right, res), 0);
        int sum = l + r + root->val;
        res = max(sum, res);
        return max(l, r) + root->val;
    }
};

File: src/com/train/algorithm/tree/implementInPython/BinaryTreeMaximumPathSum.py
import sys


class TreeNode(object):
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution(object):
    def _max(self, root):
        if not root: return -sys.maxint
        l = max(0, self._max(root.left))
        r = max(0, self._max(root.right))
        self.res = max(self.res, root.val + l + r)
        return root.val + max(l, r)
    def maxPathSum(self, root):
        self.res = -sys.maxint
        self._max(root)
        return self.res